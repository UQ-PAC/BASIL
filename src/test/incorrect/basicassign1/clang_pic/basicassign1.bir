000005b0: program
0000057e: sub __cxa_finalize(__cxa_finalize_result)
000005b1: __cxa_finalize_result :: out u32 = low:32[R0]

00000396:
0000048e: R16 := 0x11000
00000495: R17 := mem[R16 + 8, el]:u64
0000049b: R16 := R16 + 8
000004a0: call R17 with noreturn

0000057f: sub __do_global_dtors_aux(__do_global_dtors_aux_result)
000005b2: __do_global_dtors_aux_result :: out u32 = low:32[R0]

00000292:
00000296: #3 := R31 - 0x20
0000029c: mem := mem with [#3, el]:u64 <- R29
000002a2: mem := mem with [#3 + 8, el]:u64 <- R30
000002a6: R31 := #3
000002ac: R29 := R31
000002b4: mem := mem with [R31 + 0x10, el]:u64 <- R19
000002b9: R19 := 0x11000
000002c0: R0 := pad:64[mem[R19 + 0x30]]
000002c7: when 31:0[R0] <> 0 goto %000002c5
000005a6: goto %0000035f

0000035f:
00000362: R0 := 0x10000
00000369: R0 := mem[R0 + 0xFB8, el]:u64
0000036f: when R0 = 0 goto %0000036d
000005a7: goto %00000386

00000386:
00000389: R0 := 0x11000
00000390: R0 := mem[R0 + 0x28, el]:u64
00000395: R30 := 0x730
00000398: call @__cxa_finalize with return %0000036d

0000036d:
00000375: R30 := 0x734
00000377: call @deregister_tm_clones with return %00000379

00000379:
0000037c: R0 := 1
00000384: mem := mem with [R19 + 0x30] <- 7:0[R0]
000005a8: goto %000002c5

000002c5:
000002cf: R19 := mem[R31 + 0x10, el]:u64
000002d6: R29 := mem[R31, el]:u64
000002db: R30 := mem[R31 + 8, el]:u64
000002df: R31 := R31 + 0x20
000002e4: call R30 with noreturn

00000583: sub __libc_start_main(__libc_start_main_main, __libc_start_main_arg2, __libc_start_main_arg3, __libc_start_main_auxv, __libc_start_main_result)
000005b3: __libc_start_main_main :: in u64 = R0
000005b4: __libc_start_main_arg2 :: in u32 = low:32[R1]
000005b5: __libc_start_main_arg3 :: in out u64 = R2
000005b6: __libc_start_main_auxv :: in out u64 = R3
000005b7: __libc_start_main_result :: out u32 = low:32[R0]

000001eb:
00000478: R16 := 0x11000
0000047f: R17 := mem[R16, el]:u64
00000485: R16 := R16
0000048a: call R17 with noreturn

00000584: sub _fini(_fini_result)
000005b8: _fini_result :: out u32 = low:32[R0]

0000001f:
00000025: #0 := R31 - 0x10
0000002b: mem := mem with [#0, el]:u64 <- R29
00000031: mem := mem with [#0 + 8, el]:u64 <- R30
00000035: R31 := #0
0000003b: R29 := R31
00000042: R29 := mem[R31, el]:u64
00000047: R30 := mem[R31 + 8, el]:u64
0000004b: R31 := R31 + 0x10
00000050: call R30 with noreturn

00000585: sub _init(_init_result)
000005b9: _init_result :: out u32 = low:32[R0]

000004e8:
000004ee: #5 := R31 - 0x10
000004f4: mem := mem with [#5, el]:u64 <- R29
000004fa: mem := mem with [#5 + 8, el]:u64 <- R30
000004fe: R31 := #5
00000504: R29 := R31
00000509: R30 := 0x5C0
0000050b: call @call_weak_fn with return %0000050d

0000050d:
00000512: R29 := mem[R31, el]:u64
00000517: R30 := mem[R31 + 8, el]:u64
0000051b: R31 := R31 + 0x10
00000520: call R30 with noreturn

00000586: sub _start(_start_result)
000005ba: _start_result :: out u32 = low:32[R0]

000001ac:
000001b1: R29 := 0
000001b6: R30 := 0
000001bc: R5 := R0
000001c3: R1 := mem[R31, el]:u64
000001c9: R2 := R31 + 8
000001cf: R6 := R31
000001d4: R0 := 0x10000
000001db: R0 := mem[R0 + 0xFD8, el]:u64
000001e0: R3 := 0
000001e5: R4 := 0
000001ea: R30 := 0x670
000001ed: call @__libc_start_main with return %000001ef

000001ef:
000001f2: R30 := 0x674
000001f5: call @abort with return %000005a9

000005a9:
000005aa: call @call_weak_fn with noreturn

00000589: sub abort()


000001f3:
000004ba: R16 := 0x11000
000004c1: R17 := mem[R16 + 0x18, el]:u64
000004c7: R16 := R16 + 0x18
000004cc: call R17 with noreturn

0000058a: sub call_weak_fn(call_weak_fn_result)
000005bb: call_weak_fn_result :: out u32 = low:32[R0]

000001f7:
000001fa: R0 := 0x10000
00000201: R0 := mem[R0 + 0xFC8, el]:u64
00000207: when R0 = 0 goto %00000205
000005ab: goto %000003d6

00000205:
0000020d: call R30 with noreturn

000003d6:
000003d9: goto @__gmon_start__

000003d7:
000004a4: R16 := 0x11000
000004ab: R17 := mem[R16 + 0x10, el]:u64
000004b1: R16 := R16 + 0x10
000004b6: call R17 with noreturn

0000058c: sub deregister_tm_clones(deregister_tm_clones_result)
000005bc: deregister_tm_clones_result :: out u32 = low:32[R0]

00000213:
00000216: R0 := 0x11000
0000021c: R0 := R0 + 0x30
00000221: R1 := 0x11000
00000227: R1 := R1 + 0x30
0000022d: #1 := ~R0
00000232: #2 := R1 + ~R0
00000238: VF := extend:65[#2 + 1] <> extend:65[R1] + extend:65[#1] + 1
0000023e: CF := pad:65[#2 + 1] <> pad:65[R1] + pad:65[#1] + 1
00000242: ZF := #2 + 1 = 0
00000246: NF := 63:63[#2 + 1]
0000024c: when ZF goto %0000024a
000005ac: goto %000003b8

000003b8:
000003bb: R1 := 0x10000
000003c2: R1 := mem[R1 + 0xFB0, el]:u64
000003c7: when R1 = 0 goto %0000024a
000005ad: goto %000003cb

0000024a:
00000252: call R30 with noreturn

000003cb:
000003cf: R16 := R1
000003d4: call R16 with noreturn

0000058f: sub frame_dummy(frame_dummy_result)
000005bd: frame_dummy_result :: out u32 = low:32[R0]

000002ea:
000002ec: call @register_tm_clones with noreturn

00000590: sub main(main_argc, main_argv, main_result)
000005be: main_argc :: in u32 = low:32[R0]
000005bf: main_argv :: in out u64 = R1
000005c0: main_result :: out u32 = low:32[R0]

000002ee:
000002f2: R31 := R31 - 0x10
000002f7: R8 := 0x10000
000002fe: R8 := mem[R8 + 0xFD0, el]:u64
00000305: R9 := pad:64[mem[R8, el]:u32]
0000030d: mem := mem with [R31 + 0xC, el]:u32 <- 31:0[R9]
00000312: R0 := 0
00000319: mem := mem with [R31 + 0xC, el]:u32 <- 0
00000320: R10 := pad:64[mem[R31 + 0xC, el]:u32]
00000325: R9 := 0x10000
0000032c: R9 := mem[R9 + 0xFC0, el]:u64
00000334: mem := mem with [R9, el]:u32 <- 31:0[R10]
0000033b: R8 := pad:64[mem[R8, el]:u32]
00000343: mem := mem with [R31 + 0xC, el]:u32 <- 31:0[R8]
0000034a: R8 := pad:64[mem[R31 + 0xC, el]:u32]
00000352: mem := mem with [R9, el]:u32 <- 31:0[R8]
00000358: R31 := R31 + 0x10
0000035d: call R30 with noreturn

00000591: sub register_tm_clones(register_tm_clones_result)
000005c1: register_tm_clones_result :: out u32 = low:32[R0]

00000254:
00000257: R0 := 0x11000
0000025d: R0 := R0 + 0x30
00000262: R1 := 0x11000
00000268: R1 := R1 + 0x30
0000026f: R1 := R1 + ~R0 + 1
00000275: R2 := 0.63:63[R1]
0000027c: R1 := R2 + (R1 ~>> 3)
00000282: R1 := extend:64[63:1[R1]]
00000288: when R1 = 0 goto %00000286
000005ae: goto %0000039a

0000039a:
0000039d: R2 := 0x10000
000003a4: R2 := mem[R2 + 0xFE0, el]:u64
000003a9: when R2 = 0 goto %00000286
000005af: goto %000003ad

00000286:
0000028e: call R30 with noreturn

000003ad:
000003b1: R16 := R2
000003b6: call R16 with noreturn
