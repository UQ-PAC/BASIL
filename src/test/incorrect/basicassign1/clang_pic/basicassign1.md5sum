e66b2935bce8ee495962c58d43cf2443  incorrect/basicassign1/clang_pic/a.out
3712947f7951e66fba538821e87200a9  incorrect/basicassign1/clang_pic/basicassign1.adt
9075a418c7e1dc3b59b3b652ca5329bd  incorrect/basicassign1/clang_pic/basicassign1.bir
f49a8bb41a3bf145b5889c87299a578b  incorrect/basicassign1/clang_pic/basicassign1.relf
c7351a4786bff50138f4b088a09d0c95  incorrect/basicassign1/clang_pic/basicassign1.gts
