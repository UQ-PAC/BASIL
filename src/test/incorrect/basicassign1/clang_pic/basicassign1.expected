var Gamma_mem: [bv64]bool;
var Gamma_stack: [bv64]bool;
var mem: [bv64]bv8;
var stack: [bv64]bv8;
const $z_addr: bv64;
axiom ($z_addr == 69688bv64);
function L(memory: [bv64]bv8, index: bv64) returns (bool) {
  (if (index == $z_addr) then true else false)
}

function {:bvbuiltin "bvadd"} bvadd64(bv64, bv64) returns (bv64);
function gamma_load32(gammaMap: [bv64]bool, index: bv64) returns (bool) {
  (gammaMap[bvadd64(index, 3bv64)] && (gammaMap[bvadd64(index, 2bv64)] && (gammaMap[bvadd64(index, 1bv64)] && gammaMap[index])))
}

function gamma_load64(gammaMap: [bv64]bool, index: bv64) returns (bool) {
  (gammaMap[bvadd64(index, 7bv64)] && (gammaMap[bvadd64(index, 6bv64)] && (gammaMap[bvadd64(index, 5bv64)] && (gammaMap[bvadd64(index, 4bv64)] && (gammaMap[bvadd64(index, 3bv64)] && (gammaMap[bvadd64(index, 2bv64)] && (gammaMap[bvadd64(index, 1bv64)] && gammaMap[index])))))))
}

function gamma_store32(gammaMap: [bv64]bool, index: bv64, value: bool) returns ([bv64]bool) {
  gammaMap[index := value][bvadd64(index, 1bv64) := value][bvadd64(index, 2bv64) := value][bvadd64(index, 3bv64) := value]
}

function memory_load32_le(memory: [bv64]bv8, index: bv64) returns (bv32) {
  (memory[bvadd64(index, 3bv64)] ++ (memory[bvadd64(index, 2bv64)] ++ (memory[bvadd64(index, 1bv64)] ++ memory[index])))
}

function memory_load64_le(memory: [bv64]bv8, index: bv64) returns (bv64) {
  (memory[bvadd64(index, 7bv64)] ++ (memory[bvadd64(index, 6bv64)] ++ (memory[bvadd64(index, 5bv64)] ++ (memory[bvadd64(index, 4bv64)] ++ (memory[bvadd64(index, 3bv64)] ++ (memory[bvadd64(index, 2bv64)] ++ (memory[bvadd64(index, 1bv64)] ++ memory[index])))))))
}

function memory_load8_le(memory: [bv64]bv8, index: bv64) returns (bv8) {
  memory[index]
}

function memory_store32_le(memory: [bv64]bv8, index: bv64, value: bv32) returns ([bv64]bv8) {
  memory[index := value[8:0]][bvadd64(index, 1bv64) := value[16:8]][bvadd64(index, 2bv64) := value[24:16]][bvadd64(index, 3bv64) := value[32:24]]
}

function {:bvbuiltin "zero_extend 32"} zero_extend32_32(bv32) returns (bv64);
procedure rely();
  modifies mem, Gamma_mem;
  ensures (forall i: bv64 :: (((mem[i] == old(mem[i])) ==> (Gamma_mem[i] == old(Gamma_mem[i])))));
  ensures (mem == old(mem));
  ensures (Gamma_mem == old(Gamma_mem));

procedure rely_transitive()
  modifies mem, Gamma_mem;
  ensures (mem == old(mem));
  ensures (Gamma_mem == old(Gamma_mem));
{
  call rely();
  call rely();
}

procedure rely_reflexive();

procedure guarantee_reflexive();
  modifies mem, Gamma_mem;

procedure main(main_argc: bv32, Gamma_main_argc: bool, main_argv: bv64, Gamma_main_argv: bool, FP: bv64, Gamma_FP: bool, LR: bv64, Gamma_LR: bool, SP: bv64, Gamma_SP: bool) returns (main_argv_out: bv64, Gamma_main_argv_out: bool, main_result: bv32, Gamma_main_result: bool, FP_out: bv64, Gamma_FP_out: bool, LR_out: bv64, Gamma_LR_out: bool, SP_out: bv64, Gamma_SP_out: bool)
  modifies mem, Gamma_mem, stack, Gamma_stack;
  requires (memory_load8_le(mem, 1964bv64) == 1bv8);
  requires (memory_load8_le(mem, 1965bv64) == 0bv8);
  requires (memory_load8_le(mem, 1966bv64) == 2bv8);
  requires (memory_load8_le(mem, 1967bv64) == 0bv8);
  requires (memory_load8_le(mem, 69544bv64) == 200bv8);
  requires (memory_load8_le(mem, 69545bv64) == 13bv8);
  requires (memory_load8_le(mem, 69546bv64) == 1bv8);
  requires (memory_load8_le(mem, 69547bv64) == 0bv8);
  requires (memory_load8_le(mem, 69548bv64) == 0bv8);
  requires (memory_load8_le(mem, 69549bv64) == 0bv8);
  requires (memory_load8_le(mem, 69550bv64) == 0bv8);
  requires (memory_load8_le(mem, 69551bv64) == 0bv8);
  requires (memory_load8_le(mem, 69552bv64) == 0bv8);
  requires (memory_load8_le(mem, 69553bv64) == 0bv8);
  requires (memory_load8_le(mem, 69554bv64) == 0bv8);
  requires (memory_load8_le(mem, 69555bv64) == 0bv8);
  requires (memory_load8_le(mem, 69556bv64) == 0bv8);
  requires (memory_load8_le(mem, 69557bv64) == 0bv8);
  requires (memory_load8_le(mem, 69558bv64) == 0bv8);
  requires (memory_load8_le(mem, 69559bv64) == 0bv8);
  requires (memory_load8_le(mem, 69560bv64) == 0bv8);
  requires (memory_load8_le(mem, 69561bv64) == 0bv8);
  requires (memory_load8_le(mem, 69562bv64) == 0bv8);
  requires (memory_load8_le(mem, 69563bv64) == 0bv8);
  requires (memory_load8_le(mem, 69564bv64) == 0bv8);
  requires (memory_load8_le(mem, 69565bv64) == 0bv8);
  requires (memory_load8_le(mem, 69566bv64) == 0bv8);
  requires (memory_load8_le(mem, 69567bv64) == 0bv8);
  requires (memory_load8_le(mem, 69568bv64) == 56bv8);
  requires (memory_load8_le(mem, 69569bv64) == 16bv8);
  requires (memory_load8_le(mem, 69570bv64) == 1bv8);
  requires (memory_load8_le(mem, 69571bv64) == 0bv8);
  requires (memory_load8_le(mem, 69572bv64) == 0bv8);
  requires (memory_load8_le(mem, 69573bv64) == 0bv8);
  requires (memory_load8_le(mem, 69574bv64) == 0bv8);
  requires (memory_load8_le(mem, 69575bv64) == 0bv8);
  requires (memory_load8_le(mem, 69576bv64) == 0bv8);
  requires (memory_load8_le(mem, 69577bv64) == 0bv8);
  requires (memory_load8_le(mem, 69578bv64) == 0bv8);
  requires (memory_load8_le(mem, 69579bv64) == 0bv8);
  requires (memory_load8_le(mem, 69580bv64) == 0bv8);
  requires (memory_load8_le(mem, 69581bv64) == 0bv8);
  requires (memory_load8_le(mem, 69582bv64) == 0bv8);
  requires (memory_load8_le(mem, 69583bv64) == 0bv8);
  requires (memory_load8_le(mem, 69584bv64) == 52bv8);
  requires (memory_load8_le(mem, 69585bv64) == 16bv8);
  requires (memory_load8_le(mem, 69586bv64) == 1bv8);
  requires (memory_load8_le(mem, 69587bv64) == 0bv8);
  requires (memory_load8_le(mem, 69588bv64) == 0bv8);
  requires (memory_load8_le(mem, 69589bv64) == 0bv8);
  requires (memory_load8_le(mem, 69590bv64) == 0bv8);
  requires (memory_load8_le(mem, 69591bv64) == 0bv8);
  requires (memory_load8_le(mem, 69592bv64) == 84bv8);
  requires (memory_load8_le(mem, 69593bv64) == 7bv8);
  requires (memory_load8_le(mem, 69594bv64) == 0bv8);
  requires (memory_load8_le(mem, 69595bv64) == 0bv8);
  requires (memory_load8_le(mem, 69596bv64) == 0bv8);
  requires (memory_load8_le(mem, 69597bv64) == 0bv8);
  requires (memory_load8_le(mem, 69598bv64) == 0bv8);
  requires (memory_load8_le(mem, 69599bv64) == 0bv8);
  requires (memory_load8_le(mem, 69600bv64) == 0bv8);
  requires (memory_load8_le(mem, 69601bv64) == 0bv8);
  requires (memory_load8_le(mem, 69602bv64) == 0bv8);
  requires (memory_load8_le(mem, 69603bv64) == 0bv8);
  requires (memory_load8_le(mem, 69604bv64) == 0bv8);
  requires (memory_load8_le(mem, 69605bv64) == 0bv8);
  requires (memory_load8_le(mem, 69606bv64) == 0bv8);
  requires (memory_load8_le(mem, 69607bv64) == 0bv8);
  requires (memory_load8_le(mem, 69664bv64) == 0bv8);
  requires (memory_load8_le(mem, 69665bv64) == 0bv8);
  requires (memory_load8_le(mem, 69666bv64) == 0bv8);
  requires (memory_load8_le(mem, 69667bv64) == 0bv8);
  requires (memory_load8_le(mem, 69668bv64) == 0bv8);
  requires (memory_load8_le(mem, 69669bv64) == 0bv8);
  requires (memory_load8_le(mem, 69670bv64) == 0bv8);
  requires (memory_load8_le(mem, 69671bv64) == 0bv8);
  requires (memory_load8_le(mem, 69672bv64) == 40bv8);
  requires (memory_load8_le(mem, 69673bv64) == 16bv8);
  requires (memory_load8_le(mem, 69674bv64) == 1bv8);
  requires (memory_load8_le(mem, 69675bv64) == 0bv8);
  requires (memory_load8_le(mem, 69676bv64) == 0bv8);
  requires (memory_load8_le(mem, 69677bv64) == 0bv8);
  requires (memory_load8_le(mem, 69678bv64) == 0bv8);
  requires (memory_load8_le(mem, 69679bv64) == 0bv8);
{
  var Gamma_R0: bool;
  var Gamma_R1: bool;
  var Gamma_R10: bool;
  var Gamma_R29: bool;
  var Gamma_R30: bool;
  var Gamma_R31: bool;
  var Gamma_R8: bool;
  var Gamma_R9: bool;
  var R0: bv64;
  var R1: bv64;
  var R10: bv64;
  var R29: bv64;
  var R30: bv64;
  var R31: bv64;
  var R8: bv64;
  var R9: bv64;
  R0, Gamma_R0 := zero_extend32_32(main_argc), Gamma_main_argc;
  R1, Gamma_R1 := main_argv, Gamma_main_argv;
  R29, Gamma_R29 := FP, Gamma_FP;
  R30, Gamma_R30 := LR, Gamma_LR;
  R31, Gamma_R31 := SP, Gamma_SP;
  lmain:
    R31, Gamma_R31 := bvadd64(R31, 18446744073709551600bv64), Gamma_R31;
    R8, Gamma_R8 := 65536bv64, true;
    call rely();
    R8, Gamma_R8 := memory_load64_le(mem, bvadd64(R8, 4048bv64)), (gamma_load64(Gamma_mem, bvadd64(R8, 4048bv64)) || L(mem, bvadd64(R8, 4048bv64)));
    call rely();
    R9, Gamma_R9 := zero_extend32_32(memory_load32_le(mem, R8)), (gamma_load32(Gamma_mem, R8) || L(mem, R8));
    stack, Gamma_stack := memory_store32_le(stack, bvadd64(R31, 12bv64), R9[32:0]), gamma_store32(Gamma_stack, bvadd64(R31, 12bv64), Gamma_R9);
    R0, Gamma_R0 := 0bv64, true;
    stack, Gamma_stack := memory_store32_le(stack, bvadd64(R31, 12bv64), 0bv32), gamma_store32(Gamma_stack, bvadd64(R31, 12bv64), true);
    R10, Gamma_R10 := zero_extend32_32(memory_load32_le(stack, bvadd64(R31, 12bv64))), gamma_load32(Gamma_stack, bvadd64(R31, 12bv64));
    R9, Gamma_R9 := 65536bv64, true;
    call rely();
    R9, Gamma_R9 := memory_load64_le(mem, bvadd64(R9, 4032bv64)), (gamma_load64(Gamma_mem, bvadd64(R9, 4032bv64)) || L(mem, bvadd64(R9, 4032bv64)));
    call rely();
    assert (L(mem, R9) ==> Gamma_R10);
    mem, Gamma_mem := memory_store32_le(mem, R9, R10[32:0]), gamma_store32(Gamma_mem, R9, Gamma_R10);
    call rely();
    R8, Gamma_R8 := zero_extend32_32(memory_load32_le(mem, R8)), (gamma_load32(Gamma_mem, R8) || L(mem, R8));
    stack, Gamma_stack := memory_store32_le(stack, bvadd64(R31, 12bv64), R8[32:0]), gamma_store32(Gamma_stack, bvadd64(R31, 12bv64), Gamma_R8);
    R8, Gamma_R8 := zero_extend32_32(memory_load32_le(stack, bvadd64(R31, 12bv64))), gamma_load32(Gamma_stack, bvadd64(R31, 12bv64));
    call rely();
    assert (L(mem, R9) ==> Gamma_R8);
    mem, Gamma_mem := memory_store32_le(mem, R9, R8[32:0]), gamma_store32(Gamma_mem, R9, Gamma_R8);
    R31, Gamma_R31 := bvadd64(R31, 16bv64), Gamma_R31;
    main_argv_out, Gamma_main_argv_out := R1, Gamma_R1;
    main_result, Gamma_main_result := R0[32:0], Gamma_R0;
    FP_out, Gamma_FP_out := R29, Gamma_R29;
    LR_out, Gamma_LR_out := R30, Gamma_R30;
    SP_out, Gamma_SP_out := R31, Gamma_R31;
    return;
}
